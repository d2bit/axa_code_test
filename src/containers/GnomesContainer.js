import React from 'react'
import DataDownloader from '../libs/dataDownloader'
import GnomeList from '../components/GnomeList'
import GnomeFilter from '../components/GnomeFilter'

class GnomesContainer extends React.Component {
  state = {
    population: [],
    filter: {
      name: '',
    },
  }

  async componentDidMount() {
    const globalPopulation = await DataDownloader.get()
    if (!globalPopulation) return

    const population = Object.values(globalPopulation)[0]
    this.setState({ population })
  }

  updateFilter = (filter) => {
    this.setState({ filter })
  }

  filteredPopulation = () => {
    const { population, filter } = this.state
    let filtered = population.filter(gnome =>
      gnome.name.toLowerCase().includes(filter.name.toLowerCase())
    )
    return filtered
  }

  render() {
    return (
      <React.Fragment>
        <GnomeFilter updater={this.updateFilter} />
        <GnomeList population={this.filteredPopulation()} />
      </React.Fragment>
    )
  }
}

export default GnomesContainer
