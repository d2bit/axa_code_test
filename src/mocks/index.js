import data from './data.json'

const population = Object.values(data)[0]
const populationCount = population.length

export function getPopulation(size=5) {
  if (size >= populationCount) {
    return population
  }
  return population.slice(0, size)
}

export function getGnome(index) {
  if (index && index >= 0 && index < populationCount) {
    return population[index]
  }
  const randomIndex = Math.floor(Math.random() * populationCount)
  return population[randomIndex]
}

export function getResponseData() {
  return data
}

const Helper = {
  getPopulation,
  getGnome,
  getResponseData,
}

export default Helper
