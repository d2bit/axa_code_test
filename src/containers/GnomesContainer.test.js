import React from 'react'
import { render, Simulate, wait } from 'react-testing-library'
import 'dom-testing-library/extend-expect'
import GnomesContainer from './GnomesContainer'
import { getGnome } from '../mocks'
import DataDownloader from '../libs/dataDownloader'
jest.mock('../libs/dataDownloader')

describe('GnomesContainer', () => {
  DataDownloader.get.mockResolvedValue({
    town: [
      getGnome(1, { name: 'Lola' }),
      getGnome(2, { name: 'Lucia' }),
      getGnome(3, { name: 'Paco' }),
      getGnome(4, { name: 'Nestor' }),
    ]
  })

  it('fetches the population', () => {
    render(<GnomesContainer />)

    expect(DataDownloader.get).toHaveBeenCalledTimes(1)
  })

  it('filters the population', async () => {
    const { getByTestId, container } = render(<GnomesContainer />)

    await wait()
    expect(container).toMatchSnapshot()
    expect(getByTestId('Gnome-1')).toHaveTextContent('Lola')
    const gnomeFilter = getByTestId('GnomeFilter')
    Simulate.change(gnomeFilter, { target: { value: 'P' } })
    await wait()
    expect(container).not.toHaveTextContent('Lola')
    expect(container).toMatchSnapshot()
  })
})
