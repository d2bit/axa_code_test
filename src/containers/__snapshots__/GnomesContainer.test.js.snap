// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GnomesContainer filters the population 1`] = `
<div>
  <div>
    <input
      data-testid="GnomeFilter"
      name="GnomeFilter"
      value=""
    />
  </div>
  <div
    data-testid="GnomeList"
  >
    <div
      data-testid="Gnome-1"
    >
      <h3
        data-testid="name"
      >
        Lola
      </h3>
      <img
        alt="Lola"
        data-testid="thumbnail"
        src="http://www.publicdomainpictures.net/pictures/120000/nahled/white-hen.jpg"
        width="200px"
      />
    </div>
    <div
      data-testid="Gnome-2"
    >
      <h3
        data-testid="name"
      >
        Lucia
      </h3>
      <img
        alt="Lucia"
        data-testid="thumbnail"
        src="http://www.publicdomainpictures.net/pictures/30000/nahled/maple-leaves-background.jpg"
        width="200px"
      />
    </div>
    <div
      data-testid="Gnome-3"
    >
      <h3
        data-testid="name"
      >
        Paco
      </h3>
      <img
        alt="Paco"
        data-testid="thumbnail"
        src="http://www.publicdomainpictures.net/pictures/10000/nahled/1-1275919724d1Oh.jpg"
        width="200px"
      />
    </div>
    <div
      data-testid="Gnome-4"
    >
      <h3
        data-testid="name"
      >
        Nestor
      </h3>
      <img
        alt="Nestor"
        data-testid="thumbnail"
        src="http://www.publicdomainpictures.net/pictures/10000/nahled/zebra-head-11281366876AZ3M.jpg"
        width="200px"
      />
    </div>
  </div>
</div>
`;

exports[`GnomesContainer filters the population 2`] = `
<div>
  <div>
    <input
      data-testid="GnomeFilter"
      name="GnomeFilter"
      value="P"
    />
  </div>
  <div
    data-testid="GnomeList"
  >
    <div
      data-testid="Gnome-3"
    >
      <h3
        data-testid="name"
      >
        Paco
      </h3>
      <img
        alt="Paco"
        data-testid="thumbnail"
        src="http://www.publicdomainpictures.net/pictures/10000/nahled/1-1275919724d1Oh.jpg"
        width="200px"
      />
    </div>
  </div>
</div>
`;
